stan/math/prim/mat/prob/neg_binomial_2_log_glm_lpmf.hpp

```diff
@@ -0,0 +1,195 @@
+      check_consistent_sizes(function,
+                             "Rows in matrix of independent variables",
+                             x.col(0), "Vector of failure counts",  phi);
+      check_consistent_sizes(function,
+                             "Columns in matrix of independent variables",
+                             x.row(0), "Weight vector",  beta);
+
+      if (!include_summand<propto, T_x, T_beta, T_alpha, T_precision>::value)
+        return 0.0;
+
+      const size_t N = x.col(0).size();
+      const size_t M = x.row(0).size();
+
+      Array<T_partials_return, Dynamic, 1> n_arr(N, 1);
+      Array<T_partials_return, Dynamic, 1> phi_arr(N, 1);
+      {
+        scalar_seq_view<T_n> n_vec(n);
+        scalar_seq_view<T_precision> phi_vec(phi);
+        for (size_t n = 0; n < N; ++n) {
+          n_arr[n] = n_vec[n];
+          phi_arr[n] = value_of(phi_vec[n]);
+        }
+      }
+
+      Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
+      {
+        scalar_seq_view<T_beta> beta_vec(beta);
+        for (size_t m = 0; m < M; ++m) {
+          beta_dbl[m] = value_of(beta_vec[m]);
+        }
+      }
+      Matrix<T_partials_return, Dynamic, Dynamic> x_dbl = value_of(x);
+
+      Array<T_partials_return, Dynamic, 1> theta_dbl = (x_dbl * beta_dbl
+      + Matrix<double, Dynamic, 1>::Ones(N, 1) * value_of(alpha)).array();
+      Array<T_partials_return, Dynamic, 1> log_phi = phi_arr.log();
+      Array<T_partials_return, Dynamic, 1> logsumexp_eta_logphi =
+        theta_dbl.binaryExpr(log_phi,
+                             [](T_partials_return xx, T_partials_return yy) {
+                               return log_sum_exp(xx, yy);
+                             });
+      Array<T_partials_return, Dynamic, 1> n_plus_phi = n_arr + phi_arr;
+
+      // Compute the log-density.
+      if (include_summand<propto>::value) {
+        logp -= (n_arr
+          + Array<double, Dynamic, 1>::Ones(N, 1))
+          .unaryExpr([](T_partials_return xx) {
+                       return lgamma(xx);
+                     }).sum();
+      }
+      if (include_summand<propto, T_precision>::value) {
+        logp += (phi_arr.binaryExpr(phi_arr, [](T_partials_return xx,
+                                                T_partials_return yy) {
+                                                  return multiply_log(xx, yy);
+                                              })
+                - phi_arr.unaryExpr([](T_partials_return xx) {
+                                      return lgamma(xx);
+                                    })).sum();
+      }
+      if (include_summand<propto, T_x, T_beta, T_alpha, T_precision>::value)
+        logp -= (n_plus_phi * logsumexp_eta_logphi).sum();
+      if (include_summand<propto, T_x, T_beta, T_alpha>::value)
+        logp += (n_arr * theta_dbl).sum();
+      if (include_summand<propto, T_precision>::value) {
+        logp += n_plus_phi.unaryExpr([](T_partials_return xx) {
```

Comment:
Can we just pass `lgamma` in here instead of creating a lambda? I'm assuming you tried and got some type error, curious about that... 

(applies to all of these `binary`- and `unaryExpr`s)

Reply:
No, it seems we cannot just pass `lgamma`. I think `binaryExpr` and `unaryExpr` expect a lambda-abstraction (or a functor). It seems to not expect a function. Joys of C++.

Reply:
Boo.